validator/src/journal/block_manager.rs

```diff
@@ -0,0 +1,1039 @@
+                    )));
+                } else {
+                    let mut previous = None;
+
+                    for block in tail {
+                        match previous {
+                            Some(previous_block_id) => {
+                                if block.previous_block_id != previous_block_id {
+                                    return Err(BlockManagerError::MissingPredecessorInBranch(
+                                        format!(
+                                            "During Put, missing predecessor of block {}: {}",
+                                            block.header_signature, block.previous_block_id
+                                        ),
+                                    ));
+                                }
+                                previous = Some(block.header_signature.clone());
+                            }
+                            None => {
+                                if block.previous_block_id != head.header_signature {
+                                    return Err(BlockManagerError::MissingPredecessorInBranch(
+                                        format!(
+                                            "During Put, missing predecessor of block {}: {}",
+                                            block.previous_block_id, head.header_signature
+                                        ),
+                                    ));
+                                }
+
+                                previous = Some(block.header_signature.clone());
+                            }
+                        }
+                    }
+
+                    match self.block_by_block_id
+                        .get_mut(head.previous_block_id.as_str())
+                    {
+                        Some(mut tup) => {
+                            let &mut (_, ref mut internal_ref_count, _) = tup;
+                            *internal_ref_count += 1;
+                        }
+                        None => {
+                            self.anchors
+                                .increase_internal_ref_count(head.previous_block_id.as_str());
+                        }
+                    }
+                }
+            }
+            None => return Err(BlockManagerError::MissingInput),
+        }
+        let blocks_not_added_yet: Vec<Block> = branch
+            .into_iter()
+            .filter(|block| {
+                !self.block_by_block_id.contains_key(&block.header_signature)
+                    && !self.blockstore_by_name.values().any(|blockstore| {
+                        blockstore.get(vec![block.header_signature.clone()]).count() > 0
+                    })
+            })
+            .collect();
+
+        if let Some((last_block, blocks)) = blocks_not_added_yet.split_last() {
+            self.block_by_block_id.insert(
+                last_block.header_signature.clone(),
+                (last_block.clone(), 0, 0),
```

Comment:
I think this needs an external reference count of 1 to start.

Reply:
Just so I understand, the blocks that go in with an internal reference count of 0 as well as ones that have an internal reference count of 1 would all have an external reference count of 1 to start.